src/sdk/pynni/nni/_graph_utils.py

```diff
@@ -0,0 +1,446 @@
+import torch
+from torch.utils.tensorboard._pytorch_graph import NodePy, NodePyIO, NodePyOP, GraphPy
+from tensorboard.compat.proto.config_pb2 import RunMetadata
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.proto.step_stats_pb2 import StepStats, DeviceStepStats
+from tensorboard.compat.proto.versions_pb2 import VersionDef
+
+CLASSTYPE_KIND = 'ClassType'
+GETATTR_KIND = 'prim::GetAttr'
+
+_logger = logging.getLogger(__name__)
+
+def build_module_graph(model, dummy_input):
+    return TorchModuleGraph(model, dummy_input)
+
+def build_graph(model, dummy_input, verbose=False):
+    g = TorchProtoGraph(model, dummy_input, verbose)
+    return g.graph_def, g.stepstats
+
+def parse_traced_name(module_name):
+    prefix = 'TracedModule['
+    suffix = ']'
+    if module_name.startswith(prefix) and module_name.endswith(suffix):
+        module_name = module_name[len(prefix):-len(suffix)]
+    return module_name
+
+class TorchGraph:
+    """
+    This class is to extract pytorch model topology graph by tracing
+    """
+    def __init__(self, model, dummy_input):
+        """
+        Parameters
+        ----------
+        model : pytorch model
+            The model user wants to speed up
+        dummy_input : pytorch tensor
+            The dummy input for ```jit.trace```, users should put it on right device before pass in
+        """
+        assert torch.__version__ >= '1.3.1'
+
+        self.bound_model = model
+        self._trace(model, dummy_input)
+
+    def _trace(self, model, dummy_input):
+        with torch.onnx.set_training(model, False):
+            self.trace = torch.jit.trace(model, dummy_input)
+            torch._C._jit_pass_inline(self.trace.graph)
+
+class TorchProtoGraph(TorchGraph):
+    def __init__(self, model, dummy_input, verbose=False):
+        super().__init__(model, dummy_input)
+        list_of_nodes = self.parse(self.trace.graph, self.trace, dummy_input)
+        if verbose:
+            print(self.trace.graph)
+        self.stepstats = RunMetadata(step_stats=StepStats(dev_stats=[DeviceStepStats(device="/device:CPU:0")]))
+        self.graph_def = GraphDef(node=list_of_nodes, versions=VersionDef(producer=22))
+
+    def parse(self, graph, trace, args=None, omit_useless_nodes=True):
+        """This method parses an optimized PyTorch model graph and produces
+        a list of nodes and node stats for eventual conversion to TensorBoard
+        protobuf format.
+
+        Args:
+        graph (PyTorch module): The model graph to be parsed.
+        trace (PyTorch JIT TracedModule): The model trace to be parsed.
+        args (tuple): input tensor[s] for the model.
+        omit_useless_nodes (boolean): Whether to remove nodes from the graph.
+        """
+        nodes_py = GraphPy()
+        for node in graph.inputs():
+            if omit_useless_nodes:
+                if not node.uses():  # number of user of the node (= number of outputs/ fanout)
```

Comment:
number of `user`?

Reply:
This is from pytorch implementation,  means number of references.